provider/openstack/networking.go

```diff
@@ -246,3 +273,75 @@ func resolveNeutronNetwork(neutron *neutron.Client, name string) (string, error)
 	}
 	return processResolveNetworkIds(name, networkIds)
 }
+
+func makeSubnetInfo(neutron *neutron.Client, subnet neutron.SubnetV2) (network.SubnetInfo, error) {
+	_, _, err := net.ParseCIDR(subnet.Cidr)
+	if err != nil {
+		return network.SubnetInfo{}, errors.Annotatef(err, "skipping subnet %q, invalid CIDR", subnet.Cidr)
+	}
+	net, err := neutron.GetNetworkV2(subnet.NetworkId)
+	if err != nil {
+		return network.SubnetInfo{}, err
+	}
+
+	// TODO (hml) 2017-03-20:
+	// With goose updates, VLANTag can be updated to be
+	// network.segmentation_id, if network.network_type equals vlan
+	info := network.SubnetInfo{
+		CIDR:              subnet.Cidr,
+		ProviderId:        network.Id(subnet.Id),
+		VLANTag:           0,
+		AvailabilityZones: net.AvailabilityZones,
+		SpaceProviderId:   "",
+	}
+	logger.Tracef("found subnet with info %#v", info)
+	return info, nil
+}
+
+// Subnets returns basic information about the specified subnets known
+// by the provider for the specified instance or list of ids. subnetIds can be
+// empty, in which case all known are returned.
+func (n *NeutronNetworking) Subnets(instId instance.Id, subnetIds []network.Id) ([]network.SubnetInfo, error) {
+	var results []network.SubnetInfo
+	subIdSet := set.NewStrings()
+	for _, subId := range subnetIds {
+		subIdSet.Add(string(subId))
+	}
+
+	if instId != instance.UnknownId {
+		// TODO(hml): 2017-03-20
+		// Implment Subnets() for case where instId is specified
+		return nil, errors.NotImplementedf("neutron subnets with instance Id")
+	} else {
+		neutron := n.env.neutron()
+		subnets, err := neutron.ListSubnetsV2()
+		if err != nil {
+			return nil, errors.Annotatef(err, "failed to retrieve subnets")
+		}
+		if len(subnetIds) == 0 {
```

Comment:
This logic can be simplified using the functionality provided by Set. You just need to do a set difference between the set of passed in ids and those obtained by the list subnets call.